### DISCUSSION TITLE
Sort by creative menu

### DISCUSSION THREAD

**Discussion by Fiocher** [P3pp3rF1y/SophisticatedBackpacks#1062] [Ideas]
None of the existing sorting methods can be ordered like creative menu, including the latest sorting method "by mod".
e.g.
![Snipaste_2024-05-27_09-34-42](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/b89fd2c8-7270-4aa6-814c-a70bd9477d51)
![Snipaste_2024-05-27_09-33-43](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/e5328922-8e0b-4043-b468-c506dadc9b82)
![Snipaste_2024-05-27_09-33-24](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/310b038b-121a-4bf7-be55-ceacc14c7986)
![Snipaste_2024-05-27_09-33-14](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/adfb69fe-4bdc-4464-8b6d-0ab65d202711)

**Comment by P3pp3rF1y** [comment]
This is very very unlikely to make it to the mod. The reason is performance because it would require creating a sort order based on creative tabs and then using this sort order for sorting and even with caching this is likely going to have very negative effect on server when player sorts

**Reply by Fiocher** [reply]
> This is very very unlikely to make it to the mod. The reason is performance because it would require creating a sort order based on creative tabs and then using this sort order for sorting and even with caching this is likely going to have very negative effect on server when player sorts

Ok, I just think the sorting logic of this mod is unusual and seems a little confusing.
I also find some other mods to compare:

Sort by mod (Sophisticated Backpacks)
![Snipaste_2024-05-27_20-20-42](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/5dd30cf0-3753-460d-85c9-7377472cf7e5)

Sort by tags (Sophisticated Backpacks)
![Snipaste_2024-05-27_20-20-25](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/856da0b5-8f16-49ee-98c0-3daacfe942a7)

Sort (Inventory Profiles Next)
![Snipaste_2024-05-27_20-21-09](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/d3339cfd-47a0-476c-bad9-694a8a7ffde4)

JEI
![Snipaste_2024-05-27_20-22-51](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/5115f482-bff3-4da3-876e-0ef06c868af6)
![Snipaste_2024-05-27_20-23-17](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/6a821a74-1f0c-4fa2-b259-9376c4ef9ab8)
![Snipaste_2024-05-27_20-23-39](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/3509053a-acbd-4058-a447-a6e9f696641b)

Creative Menu
![Snipaste_2024-05-27_20-32-19](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/415ca3e3-d25c-4499-8ea8-22329bcaa0ef)
![Snipaste_2024-05-27_20-32-50](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/f9fe6cf4-ac45-4af8-acd9-8f35f3d121d4)
![Snipaste_2024-05-27_20-33-22](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/bbe923a6-2ab7-4a44-b7ee-842026f944a2)

**Comment by P3pp3rF1y** [comment]
The sorting is easy - sorts by name if you're sorting by item or mod name (with mod it sorts by mod first)

And IPN is a great example of sorting by creative tabs causing performance issues. They specifically had to make changes to prevent that when sorting even in smallest of inventories. And I believe there's still a switch to block that for server owners so that they can prevent players from tanking their servers by "just" sorting.

Creative tabs or JEI don't sort at all they just show items in the order the mods have initialized the items for them.

**Reply by Fiocher** [reply]
> The sorting is easy - sorts by name if you're sorting by item or mod name (with mod it sorts by mod first)
> 
> And IPN is a great example of sorting by creative tabs causing performance issues. They specifically had to make changes to prevent that when sorting even in smallest of inventories. And I believe there's still a switch to block that for server owners so that they can prevent players from tanking their servers by "just" sorting.
> 
> Creative tabs or JEI don't sort at all they just show items in the order the mods have initialized the items for them.

I seem to understand a little, if sort by the initialized items of the mod, are there any performance issues?

**Reply by P3pp3rF1y** [reply]
There most definitely are issues if you try to sort by it yes. As said creative tabs or JEI don't sort items they only show whatever mc / mods provided them for the list of item. So for example in my dev environment where I only have a very minimal list of mods added on top of sophisticated there's about 20,000 items in creative tabs so when creative tabs display they only have to get this list from mods (only for search tab because for the other tabs it's limited to specific mod), JEI caches the list so it only gets it once when you open a world and from there on just displays it. 

Now if you wanted to apply sorting against this of let's say 100 items (about what diamond single chest or iron double chest can hold so this number can get much larger) that would mean looking up 100 items in that list of 20,000, on average that would take about 100 x 10,000 or 1,000,000 operations to lookup and just meager about 700 operations to compare the order which is what it does today.
But now you also have to count in additional things like 
- bigger than very minimalistic modpacks are likely to have much larger number of items so we're speaking of probably something like 100,000 items in an average one and even more in big ones like ATM and the average number of operations would get to something like 5,000,000
- the comparisons of names done at the moment are pretty simple, but with comparisons against the sorted list you need to keep in mind that many of these items have nbt so nbt would need to be compared so those millions of operations are also much more CPU intesive
- the other thing with nbt is that the one stored in creative tab is going to be different from the one that you have in your backpack e.g. you charge the building gadget that's different nbt, you change mode on the building gadget that's different nbt, ... So in fact there may need to be even more expensive operations of partially comparing nbt when figuring out where the item is in the sort order

And with creative tabs / JEI they only initialize their list on client and I am sure player is cool with their client spending some cycles on that however sorting needs to be done on server so all of this would be pushed on server and I am sure the first thing I would see from server owners is request for setting to disable this and making it disabled by default

**Reply by Fiocher** [reply]
> There most definitely are issues if you try to sort by it yes. As said creative tabs or JEI don't sort items they only show whatever mc / mods provided them for the list of item. So for example in my dev environment where I only have a very minimal list of mods added on top of sophisticated there's about 20,000 items in creative tabs so when creative tabs display they only have to get this list from mods (only for search tab because for the other tabs it's limited to specific mod), JEI caches the list so it only gets it once when you open a world and from there on just displays it.
> 
> Now if you wanted to apply sorting against this of let's say 100 items (about what diamond single chest or iron double chest can hold so this number can get much larger) that would mean looking up 100 items in that list of 20,000, on average that would take about 100 x 10,000 or 1,000,000 operations to lookup and just meager about 700 operations to compare the order which is what it does today. But now you also have to count in additional things like
> 
> * bigger than very minimalistic modpacks are likely to have much larger number of items so we're speaking of probably something like 100,000 items in an average one and even more in big ones like ATM and the average number of operations would get to something like 5,000,000
> * the comparisons of names done at the moment are pretty simple, but with comparisons against the sorted list you need to keep in mind that many of these items have nbt so nbt would need to be compared so those millions of operations are also much more CPU intesive
> * the other thing with nbt is that the one stored in creative tab is going to be different from the one that you have in your backpack e.g. you charge the building gadget that's different nbt, you change mode on the building gadget that's different nbt, ... So in fact there may need to be even more expensive operations of partially comparing nbt when figuring out where the item is in the sort order
> 
> And with creative tabs / JEI they only initialize their list on client and I am sure player is cool with their client spending some cycles on that however sorting needs to be done on server so all of this would be pushed on server and I am sure the first thing I would see from server owners is request for setting to disable this and making it disabled by default

Thank you for your detailed explanation! 👍

When it comes to IPN, I have a question, IPN has a continuous crafting feature, which is useful for crafting items that can't be stacked, but this feature is not available on Sophisticated Backpacks.

If want to make both compatible, to whom should I create compatibility request?

![Snipaste_2024-05-28_18-16-21](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/98eadcbc-3a3f-4a57-9d7d-ce36a80d3687)
![Snipaste_2024-05-28_18-16-38](https://github.com/P3pp3rF1y/SophisticatedBackpacks/assets/82749343/6590715f-52b6-444c-83c8-a13f117fdae3)

**Comment by P3pp3rF1y** [comment]
I have that feature on the list of future features on my side, not really as a compatibility, but one that's directly included in sophisticated.

**Reply by Fiocher** [reply]
> I have that feature on the list of future features on my side, not really as a compatibility, but one that's directly included in sophisticated.

It's great. I'm looking forward to it.

**Comment by ACE1337x** [comment]
I would also really like to be able to sort by creative menu...

**Reply by P3pp3rF1y** [reply]
as explained above sorting by creative menu isn't going to happen as that would be killing to servers especially in big packs. You have a "Sort by Mod" already included which sorts by name of the mod first and then name of the item which should work pretty well for you.
